src/platform_impl/web/async.rs

```diff
@@ -0,0 +1,314 @@
+use atomic_waker::AtomicWaker;
+use once_cell::unsync::Lazy;
+use std::future;
+use std::marker::PhantomData;
+use std::ops::Deref;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::mpsc::{self, Receiver, RecvError, SendError, Sender, TryRecvError};
+use std::sync::{Arc, Condvar, Mutex, RwLock};
+use std::task::Poll;
+use wasm_bindgen::prelude::wasm_bindgen;
+use wasm_bindgen::{JsCast, JsValue};
+
+// Unsafe wrapper type that allows us to use `T` when it's not `Send` from other threads.
+// `value` **must** only be accessed on the main thread.
+pub struct MainThreadSafe<const SYNC: bool, T: 'static, E: 'static> {
```

Comment:
Since you're referencing `MainThreadSafe`, I'll note that I've since made [`MainThreadBound`](https://docs.rs/icrate/0.0.2/src/icrate/Foundation/additions/thread.rs.html#170-308), you may be able to take some inspiration there as well.

Reply:
Yeah, I've seen that! I was wondering if this thing deserves it's own crate or not ...